### PR TITLE
[WIP] JAMES-31XX Make ScrolledIterator reactive

### PR DESCRIPTION
**Pull request by blackheaven**

### CODE REVIEW

**Comment by chibenwa** [comment]
```
[f199a924da620820ea91e2976e55207648f3f057] [ERROR] /james-project/mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/search/ElasticSearchSearcher.java:[83,13] no suitable method found for flatMap(this::extr[...]omHit)
[f199a924da620820ea91e2976e55207648f3f057] [ERROR]     method reactor.core.publisher.Flux.<R>flatMap(java.util.function.Function<? super org.elasticsearch.search.SearchHit,? extends org.reactivestreams.Publisher<? extends R>>) is not applicable
[f199a924da620820ea91e2976e55207648f3f057] [ERROR]       (cannot infer type-variable(s) R
[f199a924da620820ea91e2976e55207648f3f057] [ERROR]         (argument mismatch; bad return type in method reference
[f199a924da620820ea91e2976e55207648f3f057] [ERROR]           java.util.stream.Stream<org.apache.james.mailbox.store.search.MessageSearchIndex.SearchResult> cannot be converted to org.reactivestreams.Publisher<? extends R>))
[f199a924da620820ea91e2976e55207648f3f057] [ERROR]     method reactor.core.publisher.Flux.<V>flatMap(java.util.function.Function<? super org.elasticsearch.search.SearchHit,? extends org.reactivestreams.Publisher<? extends V>>,int) is not applicable
[f199a924da620820ea91e2976e55207648f3f057] [ERROR]       (cannot infer type-variable(s) V
[f199a924da620820ea91e2976e55207648f3f057] [ERROR]         (actual and formal argument lists differ in length))
[f199a924da620820ea91e2976e55207648f3f057] [ERROR]     method reactor.core.publisher.Flux.<V>flatMap(java.util.function.Function<? super org.elasticsearch.search.SearchHit,? extends org.reactivestreams.Publisher<? extends V>>,int,int) is not applicable
[f199a924da620820ea91e2976e55207648f3f057] [ERROR]       (cannot infer type-variable(s) V
[f199a924da620820ea91e2976e55207648f3f057] [ERROR]         (actual and formal argument lists differ in length))
[f199a924da620820ea91e2976e55207648f3f057] [ERROR]     method reactor.core.publisher.Flux.<R>flatMap(java.util.function.Function<? super org.elasticsearch.search.SearchHit,? extends org.reactivestreams.Publisher<? extends R>>,java.util.function.Function<? super java.lang.Throwable,? extends org.reactivestreams.Publisher<? extends R>>,java.util.function.Supplier<? extends org.reactivestreams.Publisher<? extends R>>) is not applicable
[f199a924da620820ea91e2976e55207648f3f057] [ERROR]       (cannot infer type-variable(s) R
[f199a924da620820ea91e2976e55207648f3f057] [ERROR]         (actual and formal argument lists differ in length))
[f199a924da620820ea91e2976e55207648f3f057] [ERROR]     method reactor.core.publisher.Flux.<V>flatMap(java.util.function.Function<? super org.elasticsearch.search.SearchHit,? extends org.reactivestreams.Publisher<? extends V>>,boolean,int,int) is not applicable
[f199a924da620820ea91e2976e55207648f3f057] [ERROR]       (cannot infer type-variable(s) V
[f199a924da620820ea91e2976e55207648f3f057] [ERROR]         (actual and formal argument lists differ in length))
[f199a924da620820ea91e2976e55207648f3f057] [ERROR] -> [Help 1]
[f199a924da620820ea91e2976e55207648f3f057] [ERROR] 
[f199a924da620820ea91e2976e55207648f3f057] [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[f199a924da620820ea91e2976e55207648f3f057] [ERROR] Re-run Maven using the -X switch to enable full debug logging.
[f199a924da620820ea91e2976e55207648f3f057] [ERROR] 
[f199a924da620820ea91e2976e55207648f3f057] [ERROR] For more information about the errors and possible solutions, please read the following articles:
[f199a924da620820ea91e2976e55207648f3f057] [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[f199a924da620820ea91e2976e55207648f3f057] [ERROR] 
[f199a924da620820ea91e2976e55207648f3f057] [ERROR] After correcting the problems, you can resume the build with the command
[f199a924da620820ea91e2976e55207648f3f057] [ERROR]   mvn <args> -rf :apache-james-mailbox-elasticsearch
```

**Comment by Arsnael** [comment]
```
[156fd3532b7a58739dde4a5926a1e37fb003b5ce] [ERROR] Errors: 
[156fd3532b7a58739dde4a5926a1e37fb003b5ce] [ERROR]   ElasticSearchIndexerTest.deleteByQueryShouldWorkOnSingleMessage:164 » Reactive
[156fd3532b7a58739dde4a5926a1e37fb003b5ce] [ERROR]   ElasticSearchIndexerTest.deleteByQueryShouldWorkWhenMultipleMessages:193 » Reactive
[156fd3532b7a58739dde4a5926a1e37fb003b5ce] [ERROR]   ScrolledSearchTest.scrollIterableShouldWorkWhenEmpty:84 » Reactive java.io.IOE...
[156fd3532b7a58739dde4a5926a1e37fb003b5ce] [ERROR]   ScrolledSearchTest.scrollIterableShouldWorkWhenMoreThanSizeElement:173 » Reactive
[156fd3532b7a58739dde4a5926a1e37fb003b5ce] [ERROR]   ScrolledSearchTest.scrollIterableShouldWorkWhenOneElement:106 » Reactive java....
[156fd3532b7a58739dde4a5926a1e37fb003b5ce] [ERROR]   ScrolledSearchTest.scrollIterableShouldWorkWhenSizeElement:136 » Reactive java...
```

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/search/ScrolledSearch.java [hunk 1]
```diff
@@ -39,74 +35,74 @@
 
 import com.github.fge.lambdas.Throwing;
 
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
 public class ScrolledSearch {
```
Comment (mbaechler) [inline backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/search/ScrolledSearch.java]: you should try to implement ScrolledSearch with Flux.create(sink). It's much easier than using Futures to close the gap between async code and reactor.

Comment (blackheaven) [reply backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/search/ScrolledSearch.java]: I looked more complex when I did

**Comment by chibenwa** [comment]
Handled as part of a8dcf2192885b4c523cca85d69b06194a98301fe `JAMES-3144 Reactify ScrollSearch`

**Pull request closed without merging**